tools/apidiff/cmd/packages.go

```diff
@@ -0,0 +1,225 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/tools/apidiff/exports"
+	"github.com/Azure/azure-sdk-for-go/tools/apidiff/repo"
+	"github.com/spf13/cobra"
+)
+
+var packagesCmd = &cobra.Command{
+	Use:   "packages [package search dir] [base commit] [target commit]",
+	Short: "Generates report for all packages under the specified directory.",
+	Long: `The packages command generates a report for all of the packages under the directory specified in [package dir].
+The package content in [target commit] is compared against the package content in [base commit]
+to determine what changes were introduced in [target commit].`,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		rootDir, cloneRepo, cleanupFn, err := processArgsAndClone(args)
+		if err != nil {
+			return err
+		}
+		defer cleanupFn()
+
+		baseCommit := args[1]
+		targetCommit := args[2]
+
+		// get for lhs
+		vprintf("checking out base commit %s and gathering exports\n", baseCommit)
+		lhs, err := getRepoContentForCommit(cloneRepo, rootDir, baseCommit)
+		if err != nil {
+			return err
+		}
+
+		// get for rhs
+		vprintf("checking out target commit %s and gathering exports\n", targetCommit)
+		rhs, err := getRepoContentForCommit(cloneRepo, rootDir, targetCommit)
+		if err != nil {
+			return err
+		}
+
+		// diff and report
+		return printReport(getPkgsReport(lhs, rhs))
+	},
+}
+
+func init() {
+	rootCmd.AddCommand(packagesCmd)
+}
+
+func getRepoContentForCommit(wt repo.WorkingTree, dir, commit string) (r repoContent, err error) {
+	err = wt.Checkout(commit)
+	if err != nil {
+		err = fmt.Errorf("failed to check out commit '%s': %s", commit, err)
+		return
+	}
+
+	pkgDirs := []string{}
+	err = filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			// check if leaf dir
+			fi, err := ioutil.ReadDir(path)
+			if err != nil {
+				return err
+			}
+			hasSubDirs := false
+			for _, f := range fi {
+				if f.IsDir() {
+					hasSubDirs = true
+					break
+				}
+			}
+			if !hasSubDirs {
+				pkgDirs = append(pkgDirs, path)
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return
+	}
+	if verboseFlag {
+		fmt.Println("found the following package directories")
+		for _, d := range pkgDirs {
+			fmt.Printf("\t%s\n", d)
+		}
+	}
+
+	r, err = getExportsForPackages(pkgDirs)
+	if err != nil {
+		err = fmt.Errorf("failed to get exports for commit '%s': %s", commit, err)
+	}
+	return
+}
+
+// contains repo content, it's structured as "package":"apiversion":content
+type repoContent map[string]map[string]exports.Content
```

Comment:
>map[string]map[string]exports.Content [](start = 17, length = 37)

cant we have just the package path here and the content ? (i.e. map[string]exports.Content ) or is it helpful later ?

Reply:
It's used to group packages by API version.